packages/language-service/ivy/completions.ts

```diff
@@ -0,0 +1,250 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {AST, ImplicitReceiver, MethodCall, PropertyRead, TmplAstNode, TmplAstReference, TmplAstTemplate, TmplAstVariable} from '@angular/compiler';
+import {NgCompiler} from '@angular/compiler-cli/src/ngtsc/core';
+import {CompletionKind, ReferenceSymbol, VariableSymbol} from '@angular/compiler-cli/src/ngtsc/typecheck/api';
+import * as ts from 'typescript';
+
+import {DisplayInfoKind, getDisplayInfo, unsafeCastDisplayInfoKindToScriptElementKind} from './display_parts';
+
+type PropertyExpressionCompletionBuilder = CompletionBuilder<PropertyRead|MethodCall>;
+
+/**
+ * Closure around the context required to perform autocompletion operations regarding a given node
+ * in the template.
+ *
+ * The generic `N` type for the template node is narrowed internally for certain operations, as not
+ * all kinds of completions are possible for every input node.
+ *
+ * @param N type of the template node in question. Certain operations are only valid for specific
+ *     node types.
+ */
+export class CompletionBuilder<N extends TmplAstNode|AST> {
+  private readonly typeChecker = this.compiler.getNextProgram().getTypeChecker();
+
+  constructor(
+      private readonly tsLS: ts.LanguageService, private readonly compiler: NgCompiler,
+      private readonly component: ts.ClassDeclaration, private readonly node: N,
+      private readonly context: TmplAstTemplate|null) {}
+
+  /**
+   * Analogue for `ts.LanguageService.getCompletionsAtPosition`.
+   */
+  getCompletionsAtPosition(options: ts.GetCompletionsAtPositionOptions|
+                           undefined): ts.WithMetadata<ts.CompletionInfo>|undefined {
+    console.error('getCompletionsAtPosition()');
+    if (this.isPropertyExpressionCompletion()) {
+      return this.getPropertyExpressionCompletion(options);
+    } else {
```

Comment:
nit, no action required: branch is unnecessary

Reply:
I meant that you could just do
```suggestion
    if (this.isPropertyExpressionCompletion()) {
      return this.getPropertyExpressionCompletion(options);
    }
    return undefined;
```

no action required, just a note